### PR TITLE
audio: Implement cubeb audio out backend.

### PR DESCRIPTION
**Pull request by squidbus**
Implements a new default audio backend using cubeb, along with a config option to select an audio backend.

This backend does not suffer from https://github.com/shadps4-emu/shadPS4/issues/1717 (at least it fixes it for me), can reduce audio stutter, and implements some extra features like partial support for `sceAudioOutSetVolume`.

### CODE REVIEW

**Comment by shinra-electric** [comment]
This works great! On Mac, it fixes audio in Bloodborne (and other souls games), as well as Full Throttle that had the same issue.

It also fixes slight audio stuttering I had in P4G and a few more games.

**Comment by rainmakerv3** [comment]
Crashes bloodborne on boot for me and a couple of other testers, seemingly no error on log
[shad_log.txt](https://github.com/user-attachments/files/18253081/shad_log.txt)

**Comment by ycb678** [comment]
When can it be integrated

**Comment by rharish101** [comment]
I can confirm that the new cubeb backend solves the audio issue for me! Switching back to the SDL backend brings back the distorted audio. I'm running an AMD CPU (7800X3D) and an AMD GPU (7900 XTX). I tested this on Wayland (Hyprland, after the `unset DISPLAY` trick) using Pipewire. I didn't test on X11 (still with Pipewire), but I was seeing the same behaviour.

**Comment by squidbus** [comment]
> Crashes bloodborne on boot for me and a couple of other testers, seemingly no error on log

Could you or anyone else testing with this issue capture in a debugger where the crash is coming from?

**Comment by GHU7924** [comment]
>Crashes bloodborne on boot for me and a couple of other testers, seemingly no error on log

Yes, it also crashed without a critical error. (Intel + Nvidia)

[shad_log.txt](https://github.com/user-attachments/files/18255274/shad_log.txt)

@squidbus 

>[Lib.Ajm] <Info> ajm.cpp:sceAjmInitialize:137: called reserved = 0
[Lib.Ajm] <Info> ajm.cpp:sceAjmModuleRegister:197: called context = 1, codec_type = 1
[Lib.AudioOut] <Info> audioout.cpp:sceAudioOutOpen:369: id = 255 port_type = MAIN index = 0 length = 256 sample_rate = 48000 param_type = FLOAT_8CH attr = NONE

After that, the crash seemed to happen. If I understood correctly. I could be wrong.

**Comment by squidbus** [comment]
Can someone with crash test again?

**Comment by Missake212** [comment]
Bloodborne still crashes on boot with no critical error 

[shad_log.txt](https://github.com/user-attachments/files/18255414/shad_log.txt)

**Comment by GHU7924** [comment]
Build 5ca0360

SDL - works fine.
cubeb - does not work.

As I understand it, the crash occurs before the logo is played.
log

[shad_log.txt](https://github.com/user-attachments/files/18255518/shad_log.txt)

EDIT: [Lib.Ajm] and [Lib.AudioOut] are not registered in the registration log.

**Comment by bigol83** [comment]
Cubeb improves framerate in Resident Evil 4 and Yakuza 0 by around 4-7 fps.

The only issue is that the first time i select a new backend it crashes the emulator, but then after that it's all good.

Yakuza 0

SDL
Immagine 2024-12-26 224817

Cubeb
Immagine 2024-12-26 224639

Resident Evil 4

SDL
sdl

Cubeb
cubeb

**Comment by squidbus** [comment]
> The only issue is that the first time i select a new backend it crashes the emulator, but then after that it's all good.

Can you get a trace of where this crash occurs?

**Comment by bigol83** [comment]
> > The only issue is that the first time i select a new backend it crashes the emulator, but then after that it's all good.
> 
> Can you get a trace of where this crash occurs?

How can i do that?

I need to correct that though. The crash doesn't happen when changing backend, but when clicking to run a game the first time. After that, it's good.

**Comment by GHU7924** [comment]
>Can you get a trace of where this crash occurs?

The first time something goes wrong, the second time you can get more.

So I launched TLG.
TLG

but if you look at the shad_log, you'll see that it's completely empty and doesn't weigh anything at all.
This seems to happen the first time any game is launched.

**Comment by squidbus** [comment]
> How can i do that?

If you can open it in a debugger like with Visual Studio it should give you a trace when the crash happens to where in the code it crashed, if not I can try to reproduce

**Comment by bigol83** [comment]
> > How can i do that?
> 
> If you can open it in a debugger like with Visual Studio it should give you a trace when the crash happens to where in the code it crashed, if not I can try to reproduce

managed to reproduce the crash while debugging in Visual Studio, it shows this

image

i checked and it only happens when cubeb is selected, when sdl is selected it doesn't happen

**Comment by squidbus** [comment]
> managed to reproduce the crash while debugging in Visual Studio, it shows this

This seems to be the crash that was first reported, was looking for the first time config crash but I guess maybe they're the same crash in the end?

**Comment by bigol83** [comment]
> > managed to reproduce the crash while debugging in Visual Studio, it shows this
> 
> This seems to be the crash that was first reported, was looking for the first time config crash but I guess maybe they're the same crash in the end?

Yeah, my bad. It is not a crash that happens during configuration, but when running a game the first time. In release build after the first time then it works without issues, while in debug build it continues to crash when trying to run a game.

But it only happens when cubeb is selected as audio backend

**Comment by rainmakerv3** [comment]
> > > How can i do that?
> > 
> > 
> > If you can open it in a debugger like with Visual Studio it should give you a trace when the crash happens to where in the code it crashed, if not I can try to reproduce
> 
> managed to reproduce the crash while debugging in Visual Studio, it shows this
> 


image


i checked and this is the same debug result as the bloodborne crash

**Comment by squidbus** [comment]
The crash should be gone now.

**Comment by TagoSMD** [comment]
BB don't crash anymore on windows

**Comment by ycb678** [comment]
Pls when will it be merged to the main version

**Comment by squidbus** [comment]
> Probably an unrelated file system inaccuracy ? loading a file with no name but just an extension is weird

Yes this is likely a game bug with loading/playing the sounds not a bug with the audio system.

File: src/common/ringbuffer.h [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+ *   other thread.
+ * - Callers are expected to provide buffers. When writing to the queue,
+ *   elements are copied into the internal storage from the buffer passed in.
+ *   When reading from the queue, the user is expected to provide a buffer.
+ *   Because this is a ring buffer, data might not be contiguous in memory,
+ *   providing an external buffer to copy into is an easy way to have linear
+ *   data for further processing.
+ */
+template <typename T>
+class ring_buffer_base {
```
Comment (raphaelthegreat) [inline src/common/ringbuffer.h]:
```suggestion
class RingBufferBase {
```
Should follow the codebase style ideally if brought into common.

**Comment by GHU7924** [comment]
Build 42fb06d (Intel + Nvidia)

Games still don't start the first time, but only the second time.

Bloodborne crash (I played for about 5 minutes. But the error does not apply to this PR, but I will attach my Log anyway.).
[shad_log.txt](https://github.com/user-attachments/files/18261964/shad_log.txt)

File: src/common/ringbuffer.h (hunk 1 above)
Comment (squidbus) [reply src/common/ringbuffer.h]: Okay, done.

**Comment by squidbus** [comment]
> Games still don't start the first time, but only the second time.

This dosn't really make sense as if it crashes the emulator restarts so there shouldn't be a difference. Please send a trace from a debugger as this should not be the one from before.

**Review by raphaelthegreat — APPROVED** [review]

**Comment by GHU7924** [comment]
>This dosn't really make sense as if it crashes the emulator restarts so there shouldn't be a difference.

https://github.com/user-attachments/assets/330fc07d-9ff0-436a-b4ef-8bd13ff5617e

This is how the first launch happens.

I tested PR 1896 and it's the same there. I do not know what could have affected it, but I do not have it in the Main. At the moment, that's for sure.

>Please send a trace from a debugger as this should not be the one from before.

I apologize. I don't have Visual Studio installed, and even if I did, I still don't know how to use it all.

**Comment by squidbus** [comment]
> I tested PR 1896 and it's the same there. I do not know what could have affected it, but I do not have it in the Main. At the moment, that's for sure.

If it's in other PRs then it's likely not caused by this one. None of the new cubeb stuff added here is even initialized at that point in launch yet.

**Comment by GHU7924** [comment]
>If it's in other PRs then it's likely not caused by this one. None of the new cubeb stuff added here is even initialized at that point in launch yet.

Apparently so, but I found out about it myself about an hour ago.
PR 1896 is also not merged and has draft status.

**Review by georgemoralis — APPROVED** [review]

**Pull request merged**